examples/flexvolume/README.md

```diff
@@ -0,0 +1,74 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/examples/flexvolume/README.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Flexvolume
+
+Flexvolume enables users to mount vendor volumes into kubernetes. It expects
+vendor drivers are installed in the volume plugin path on every kubelet node.
+
+It allows for vendors to develop their own drivers for volume plugin to attach or provision volumes on nodes.
+
+## Prerequisites
+
+Install the vendor driver on all nodes in the kubelet plugin path. Path for installing the plugin: /usr/libexec/kubernetes/kubelet-plugins/volume/exec/\<vendor~driver\>/\<driver\>
+
+For example to add a 'cifs' driver, by vendor 'foo' install the driver at: /usr/libexec/kubernetes/kubelet-plugins/volume/exec/\<foo~cifs\>/cifs
+
+## Plugin details
+
+Driver will be invoked with 'Init' to initalize the driver. It will be invoked 'setup' to attach or provision the volume on the node. It will be invoked 'teardown' to detach or destroy the volume.
```

Comment:
> @saad-ali: Builder.Setup() is only responsible for attaching a pre-existing volume and then mounting it. It should not do any automatic provisioning of volumes. Since the flex volume plugin handles the mounting part in SetUpAt(...), the 3rd party script should only be responsible for attaching a pre-existing volume in response to a setup call (nothing else). @markturansky is in the process of adding dynamic provisioning support. See #13650 and #14537. In short, the Creater interface should be the hook that results in volume provisioning.


> @nelcy Today we don't have create interface. So I made the creation part of setup for ephemeral volumes like LVM volumes. When we have a creator interface, we will extend the driver to invoke "create" to create the volume and I will make the definition explicit in docs that setup should only mount a pre-existing volume.


Today Kubernetes just doesn't support dynamic volume provisioning. I don't want to give a green flag for plugin writers to start doing it in the `init` script (as a hack essentially), because that would break the expectations the rest of the system has of this code (it attaches an existing volume).

Once dynamic provisioning is introduced, we can update this interface to add a separate creation script. For now, I'd like these docs to be very explicit that the `init` scripts is simply to attach a volume (not create it, not mount it--as an aside, maybe we should consider changing the name).